crates/bevy_sprite/src/mesh2d/mesh.rs

```diff
@@ -178,29 +184,57 @@ bitflags::bitflags! {
     }
 }
 
+pub struct RenderMesh2dInstance {
+    pub transforms: Mesh2dTransforms,
+    pub mesh_asset_id: AssetId<Mesh>,
+    pub material_bind_group_id: Material2dBindGroupId,
+    pub automatic_batching: bool,
+}
+
+#[derive(Default, Resource, Deref, DerefMut)]
+pub struct RenderMesh2dInstances(PassHashMap<Entity, RenderMesh2dInstance>);
+
+#[derive(Component)]
+pub struct Mesh2d;
+
 pub fn extract_mesh2d(
     mut commands: Commands,
-    mut previous_len: Local<usize>,
-    query: Extract<Query<(Entity, &ViewVisibility, &GlobalTransform, &Mesh2dHandle)>>,
+    mut render_mesh_instances: ResMut<RenderMesh2dInstances>,
+    query: Extract<
+        Query<(
+            Entity,
+            &ViewVisibility,
+            &GlobalTransform,
+            &Mesh2dHandle,
+            Has<NoAutomaticBatching>,
+        )>,
+    >,
 ) {
-    let mut values = Vec::with_capacity(*previous_len);
-    for (entity, view_visibility, transform, handle) in &query {
+    let capacity = query.iter().len();
+    render_mesh_instances.clear();
+    let mut entities = Vec::with_capacity(capacity);
```

Comment:
I think the `previous_len` system is better here. Consider a situation where most of the sprites are not visible.

Reply:
Done.